src/scf/self_consistent_field.jl

```diff
@@ -133,25 +91,156 @@ Solve the Kohn-Sham equations with a SCF algorithm, starting at ρ.
 
     # Measure for the accuracy of the SCF
     # TODO probably should be tracked all the way ...
     norm_Δρ = norm(info.ρout - info.ρin) * sqrt(basis.dvol)
 
     # Callback is run one last time with final state to allow callback to clean up
-    info = (; ham, basis, energies, converged, occupation_threshold,
-            ρ=ρout, α=damping, eigenvalues, occupation, εF,
-            n_iter, n_ep_extra, ψ, info.diagonalization,
-            stage=:finalize, algorithm="SCF", norm_Δρ)
+    info = (; ham, basis, energies, converged, bands.occupation_threshold,
+            ρ=ρout, α=damping, eigenvalues, occupation, εF, info.n_bands_converge,
+            n_iter, ψ, info.diagonalization, stage=:finalize,
+            algorithm="SCF", norm_Δρ)
     callback(info)
     info
 end
+
+
+"""
+Obtain new density ρ by diagonalizing `ham`. Follows the policy imposed by the `bands`
+data structure to determine and adjust the number of bands to be computed.
+"""
+function next_density(ham::Hamiltonian, bands=AdaptiveBands(ham.basis.model);
+                      eigensolver=lobpcg_hyper, ψ=nothing, eigenvalues=nothing,
+                      occupation=nothing, kwargs...)
+    n_bands_converge, n_bands_compute = determine_n_bands(bands, occupation, eigenvalues, ψ)
+
+    if isnothing(ψ)
+        increased_n_bands = true
+    else
+        @assert length(ψ) == length(ham.basis.kpoints)
+        n_bands_compute = max(n_bands_compute, maximum(ψk -> size(ψk, 2), ψ))
+        increased_n_bands = n_bands_compute > size(ψ[1], 2)
+    end
+
+    # TODO Synchronize since right now it is assumed that the same number of bands are
+    #      computed for each k-Point
+    n_bands_compute = mpi_max(n_bands_compute, ham.basis.comm_kpts)
+
+    eigres = diagonalize_all_kblocks(eigensolver, ham, n_bands_compute; ψguess=ψ,
+                                     n_conv_check=n_bands_converge, kwargs...)
+    eigres.converged || (@warn "Eigensolver not converged" iterations=eigres.iterations)
+
+    # Check maximal occupation of the unconverged bands is sensible.
+    occupation, εF = compute_occupation(ham.basis, eigres.λ)
+    minocc = maximum(minimum, occupation)
+
+    # TODO This is a bit hackish, but needed right now as we increase the number of bands
+    #      to be computed only between SCF steps. Should be revisited once we have a better
+    #      way to deal with such things in LOBPCG.
+    if !increased_n_bands && minocc > bands.occupation_threshold
+        @warn("Detected large minimal occupation $minocc. SCF could be unstable. " *
+              "Try switching to adaptive band selection (`bands=AdaptiveBands(basis)`) " *
+              "or request more converged bands than $n_bands_converge (e.g. " *
+              "`bands=AdaptiveBands(basis; n_bands_converge=$(n_bands_converge + 3)`)")
+    end
+
+    # TODO We should probably set occupation values below occupation_threshold explicitly
+    #      to zero and ignore these bands in the density computation
```

Comment:
we should not pretend they are zero if they are not, but we could pass the threshold to compute_density